File: solve/4949.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

int main()
{
    char buffer[128];
    while (1)
    {
        scanf("%[^\n]", buffer);
        if (strcmp(buffer, ".") == 0)
            break;

        int index = 0;
        int lastOpener = -1; // 1: () 2: []
        int small = 0;
        int big = 0;
        char next;
        bool result = true;
        while (buffer[index++] != '.')
        {
            if (!result)
                continue;

            if (next == '(')
            {
                // if (lastOpener == 2)
                //     result = false;
                lastOpener = 1;
                small++;
            }
            else if (next == ')')
            {
                small--;
                if (small < 0)
                    result = false;
            }
            else if (next == '[')
            {
                // if (lastOpener == 1)
                //     result = false;
                lastOpener = 2;
                big++;
            }
            else if (next == ']')
            {
                big--;
                if (big < 0)
                    result = false;
            }
        }

        bool isValid = (result && small == 0 && big == 0);

        if (isValid)
            printf("yes\n");
        else
            printf("no\n");
    }
    return 0;
}

File: solve/4949.py
while True:
    inp = input()
    if inp == ".":
        break
    
    result = True
    stk = list()
    for i in inp:
        if i == '(':
            stk.append(1)
        elif i == ')':
            if len(stk) == 0 or stk.pop() != 1:
                result = False
                break
        elif i == '[':
            stk.append(2)
        elif i == ']':
            if len(stk) == 0 or stk.pop() != 2:
                result = False
                break

    if result and len(stk) == 0:
        print("yes")
    else:
        print("no")